[rocq] Initial support for the Rocq Prover build language

ejgallego opened the pull request:
We introduce `(lang rocq)`, a build mode for the Rocq Prover, previously known as "Coq".

The build mode is an exact copy of the `(lang coq)` mode, renaming the modules, data, and some global declarations. Also, the default implicit lib is now named `Corelib`.

`(lang coq)` should be considered deprecated / frozen, and eventually removed.

There are many cleanups and improvements possible, in particular we have done, in separate commits:

- removed support for older legacy coq modes
- removed support for legacy plugin handling
- removed the need for coq-* shims

All of the tests have been ported and manually verified to work (a time consuming process) As of today, all test pass.

Future work:

- [ ] support new extraction option (-output-directory)
- [ ] improve composition with local rocq as in `(lang coq)`, as of today the rocqworker is not injected as a dependency
- [ ] nix CI + develop
- [x] forward port _RocqProject generation
- [x] forward port rocqdoc headers work
- [x] forward port per_module fixes
- [ ] cosmetic, turn `rocq` into `rocq compile` for `--display=short`
- [ ] [CI] Remove dependency on rocq-stdlib, only extraction tests need that

Closes #11572
Cc #12788 #6445 #9348 #12734 

Joint work with Li-yao Xia.

ejgallego committed:
[rocq] Initial support for the Rocq Prover build language

We introduce `(lang rocq)`, a build mode for the Rocq Prover,
previously known as "Coq".

The build mode is an exact copy of the `(lang coq)` mode, renaming the
modules, data, and some global declarations. Also, the default
implicit lib is now named `Corelib`.

(lang coq): should be considered deprecated / frozen, and eventually
removed.

There are many cleanups and improvements possible, these will be done
in later commits.

All of the tests have been ported and manually verified to work (a
time consuming process) As of today, all test pass.

The test suite has been tested against Rocq 9.1, with the `coq`
compatibility package added.

Co-authored-by: Li-yao Xia <lysxia@gmail.com>
Signed-off-by: Emilio Jesus Gallego Arias <e+git@x80.org>

ejgallego committed:
[rocq] Remove code for old Coq language versions from Rocq mode

Co-authored-by: Li-yao Xia <lysxia@gmail.com>
Signed-off-by: Emilio Jesus Gallego Arias <e+git@x80.org>

ejgallego committed:
[rocq] Documentation / reference manual.

For now a copy of Coq's one, with minor updates.

Co-authored-by: Li-yao Xia <lysxia@gmail.com>
Signed-off-by: Emilio Jesus Gallego Arias <e+git@x80.org>

ejgallego committed:
[rocq] Remove -I flags for legacy plugins

Not needed anymore as Rocq dropped support for non-findlib loading.

Co-authored-by: Li-yao Xia <lysxia@gmail.com>
Signed-off-by: Emilio Jesus Gallego Arias <e+git@x80.org>

ejgallego committed:
[rocq] Native compiler rules cleanup.

This is smaller than expected, we still need to pass the native
options and warnings due to upstream.

Co-authored-by: Li-yao Xia <lysxia@gmail.com>
Signed-off-by: Emilio Jesus Gallego Arias <e+git@x80.org>

ejgallego committed:
[rocq] Replace COQLIB and COQPATH by ROCQLIB and ROCQPATH versions.

Beware both `rocq c --config` and `coqc --config` still output `COQ*`
variable names.

Co-authored-by: Li-yao Xia <lysxia@gmail.com>
Signed-off-by: Emilio Jesus Gallego Arias <e+git@x80.org>

rlepigre-skylabs-ai commented:
I just learned that you are working on this @ejgallego, that's great news!

There are several points that are IMO worth discussing before merging this PR:
- Should we integrate the `_CoqProject` generation PR (https://github.com/ocaml/dune/pull/11752)?
- To more easily support the above, should we remove per-module flags? (IMO they are kind of useless.)
- Similarly, should we enforce at most one theory stanza per directory? (I personally always do that, can't imagine this breaking too much stuff.)
- Should we sort out the `.vos` / `.vok` mess?
- Should we integrate https://github.com/ocaml/dune/pull/11131?

Also, I'm happy to play a part in this if you want help.

ejgallego committed:
[rocq] [ci] Update to Rocq 9.1.0, don't install coq-* packages

Co-authored-by: Li-yao Xia <lysxia@gmail.com>
Signed-off-by: Emilio Jesus Gallego Arias <e+git@x80.org>

ejgallego committed:
[rocq] Changes entry

Co-authored-by: Li-yao Xia <lysxia@gmail.com>
Signed-off-by: Emilio Jesus Gallego Arias <e+git@x80.org>

ejgallego committed:
[rocq] Remove need for coqc shims.

We remove the need for installing coqc shims, and adapt the test suite
accordingly.

There is one mayor caveat and one minor caveat:

- the major caveat is that in builds composed with Rocq, dune will
  only search for `rocq`, but we don't inject a dependency on
  `rocqworker` which is also needed, but a private binary. This means
  that users should ensure manually `rocqworker` is built when using
  composed builds. We hope to fix this soon, in a different PR.

- the minor caveat is that with `--display=short` we just now display
  the `rocq` name, not `rocq compile` or `rocq dep`. This can be also
  fixed in a different PR.

Co-authored-by: Li-yao Xia <lysxia@gmail.com>
Signed-off-by: Emilio Jesus Gallego Arias <e+git@x80.org>

ejgallego committed:
[rocq] Forward port rocqdoc_headers PR #11131

Co-authored-by: Rodolphe Lepigre <lepigre@mpi-sws.org>
Signed-off-by: Emilio Jesus Gallego Arias <e+git@x80.org>

ejgallego committed:
[rocq] Foward port "Coq/Rocq: fix handling of (modules_flags ...). #12733"

Co-authored-by: Rodolphe Lepigre <lepigre@mpi-sws.org>
Signed-off-by: Emilio Jesus Gallego Arias <e+git@x80.org>

ejgallego committed:
[rocq] Forward-port "Generate "_CoqProject" files for Coq theories" #11752

Co-authored-by: Rodolphe Lepigre <lepigre@mpi-sws.org>
Signed-off-by: Emilio Jesus Gallego Arias <e+git@x80.org>

Alizter commented:
@gridbugs Do you know why the CI is failing here? The test that is affected is `portable-lockdirs-platform-dependant-dependencies.t`. AFAICT the changes in this PR are seemingly unrelated.

ejgallego committed:
[test-suite] Update portable-lockdirs test

Signed-off-by: Emilio Jesus Gallego Arias <e+git@x80.org>

Alizter commented on test/blackbox-tests/test-cases/pkg/portable-lockdirs/portable-lockdirs-platform-dependant-dependencies.t:
```diff
@@ -74,13 +74,13 @@ A package that conditionally depends on packages depending on the OS:
   $ ls $pkg_root/
-  foo.0.0.1-9b0ae1f26c9a79e6ae8bca96ec389858
-  linux-only.0.0.1-f1f7456a7bf1c70f9203f8caadf79f6d
+  foo.0.0.1-5e48eb7073ada94c09fb13ac3853f1e9
```
The fact that this had to change is worrying.

gridbugs replied:
It's because we use `Digest_feed.generic` to hash actions which contain pforms, and this change introduces a new pform constructor before the `Pkg_self` constructor, which is one of the pforms appearing in the build commands of the packages whose hashes are changing here. `Digest_feed.generic` computes a hash based on the representation of values, and variant constructors have a tag in their representation based on their position within the variant type definition. All the constructors below the new `Rocq_config` constructor have had their tag increased by 1.

To confirm this, apply the following patch which will cause the hashes in this test to revert to their original values:
```patch
diff --git i/src/dune_lang/pform.ml w/src/dune_lang/pform.ml
index e836d64c8..9b4601cc6 100644
--- i/src/dune_lang/pform.ml
+++ w/src/dune_lang/pform.ml
@@ -296,12 +296,12 @@ module Macro = struct
     | Path_no_dep
     | Ocaml_config
     | Coq_config
-    | Rocq_config
     | Env
     | Artifact of Artifact.t
     | Pkg
     | Pkg_self
     | Ppx
+    | Rocq_config
 
   let compare x y =
     match x, y with
diff --git i/src/dune_lang/pform.mli w/src/dune_lang/pform.mli
index ae2eba7c3..4feb23cf2 100644
--- i/src/dune_lang/pform.mli
+++ w/src/dune_lang/pform.mli
@@ -137,12 +137,12 @@ module Macro : sig
     | Path_no_dep
     | Ocaml_config
     | Coq_config
-    | Rocq_config
     | Env
     | Artifact of Artifact.t
     | Pkg
     | Pkg_self
     | Ppx
+    | Rocq_config
 
   val compare : t -> t -> Ordering.t
   val to_dyn : t -> Dyn.t
```

The consequence of this for users is that they'll have to rebuild many of the dependencies of projects upon updating Dune. I don't think that's a big deal since updating Dune is quite rare and I don't think that behaviour is too surprising. If we really wanted to prevent this sort of thing we'd have to stop using `Dune_feed.generic` and take care to manually implement hash functions in a way that isn't sensitive to these sorts of changes. Personally I don't think it's worth it.

ejgallego commented on src/dune_lang/pform.ml:
```diff
@@ -349,6 +350,8 @@ module Macro = struct
     | Coq_config, Coq_config -> Eq
     | Coq_config, _ -> Lt
     | _, Coq_config -> Gt
+    | Rocq_config, _ -> Lt
```
Thanks, will fix.
